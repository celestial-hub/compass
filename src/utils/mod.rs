use crate::{lexer::Lexer, parser::Parser};

pub fn ast_from_code_str(code: &str, test_name: &str) -> String {
  let ast = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
    Parser::new().parse(Lexer::new(code, test_name))
  }));

  match ast.unwrap() {
    Ok(ast) => format!("{:#?}", ast),
    Err(err) => format!("{:#?}", err),
  }
}
