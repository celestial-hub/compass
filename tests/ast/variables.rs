use compass::utils::ast_from_code_str;

#[test]
fn should_assign_i32() {
  insta::assert_snapshot!(ast_from_code_str(
    r#"a: i32 = 13"#,
    "variables/should_assign_i32/default"
  ));

  insta::assert_snapshot!(ast_from_code_str(
    r#"a: i32 = 13i32"#,
    "variables/should_assign_i32/with_type"
  ));
}

#[test]
fn should_assign_f32() {
  insta::assert_snapshot!(ast_from_code_str(
    r#"a: f32 = 12.0"#,
    "variables/should_assign_f32/default"
  ));

  insta::assert_snapshot!(ast_from_code_str(
    r#"a: f32 = 2."#,
    "variables/should_assign_f32/suffix_missing"
  ));

  insta::assert_snapshot!(ast_from_code_str(
    r#"a: f32 = .2"#,
    "variables/should_assign_f32/prefix_missing"
  ));

  insta::assert_snapshot!(ast_from_code_str(
    r#"a: f32 = 14.0f32"#,
    "variables/should_assign_f32/default_with_type"
  ));

  insta::assert_snapshot!(ast_from_code_str(
    r#"a: f32 = .3f32"#,
    "variables/should_assign_f32/prefix_missing_with_type"
  ));

  insta::assert_snapshot!(ast_from_code_str(
    r#"a: f32 = 2.f32"#,
    "variables/should_assign_f32/suffix_missing_with_type"
  ));

  insta::assert_snapshot!(ast_from_code_str(
    r#"a: f32 = 1f32"#,
    "variables/should_assign_f32/decimal_with_type"
  ));
}

#[test]
fn should_assign_f64() {
  insta::assert_snapshot!(ast_from_code_str(
    r#"a: f64 = 14.0f64"#,
    "variables/should_assign_f64/default_with_type"
  ));

  insta::assert_snapshot!(ast_from_code_str(
    r#"a: f64 = .3f64"#,
    "variables/should_assign_f64/prefix_missing_with_type"
  ));

  insta::assert_snapshot!(ast_from_code_str(
    r#"a: f64 = 2.f64"#,
    "variables/should_assign_f64/suffix_missing_with_type"
  ));

  insta::assert_snapshot!(ast_from_code_str(
    r#"a: f64 = 1f64"#,
    "variables/should_assign_f64/decimal_with_type"
  ));
}

#[test]
fn should_assign_sum_of_i32() {
  insta::assert_snapshot!(ast_from_code_str(
    r#"a: i32 = 13 + 14"#,
    "variables/should_assign_sum_of_i32/default"
  ));

  insta::assert_snapshot!(ast_from_code_str(
    r#"a: i32 = 13i32 + 14i32"#,
    "variables/should_assign_sum_of_i32/with_type"
  ));
}

#[test]
fn should_mismatch_type_i32() {
  insta::assert_snapshot!(ast_from_code_str(
    r#"a: i32 = 13.0"#,
    "variables/should_mismatch_type_i32/default"
  ));

  insta::assert_snapshot!(ast_from_code_str(
    r#"
    b: f32 = 13.0
    a: i32 = b
    "#,
    "variables/should_mismatch_type_i32/from_variable"
  ));
}

#[test]
fn should_mismatch_type_f32() {
  insta::assert_snapshot!(ast_from_code_str(
    r#"a: f32 = 13"#,
    "variables/should_mismatch_type_f32/default"
  ));

  insta::assert_snapshot!(ast_from_code_str(
    r#"
    b: i32 = 13
    a: f32 = b
    "#,
    "variables/should_mismatch_type_f32/from_variable"
  ));
}
