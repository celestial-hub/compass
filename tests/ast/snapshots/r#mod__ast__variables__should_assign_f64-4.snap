---
source: tests/ast/variables.rs
expression: "ast_from_code_str(r#\"a: f64 = 1f64\"#,\n    \"variables/should_assign_f64/decimal_with_type\")"
---
[
    Variable {
        var_type: F64,
        name: "a",
        value: LiteralF64(
            1.0,
        ),
        location: 0..0,
    },
]
